jgrapht-core/src/main/java/org/jgrapht/alg/matching/blossom/v5/KolmogorovMinimumWeightPerfectMatching.java

```diff
@@ -0,0 +1,889 @@
+ * matching at all, the algorithm doesn't find a minimum weight maximum matching, it thrown an exception.
+ *
+ * @param <V> the graph vertex type
+ * @param <E> the graph edge type
+ * @author Timofey Chudakov
+ * @see BlossomVPrimalUpdater
+ * @see BlossomVDualUpdater
+ * @since June 2018
+ */
+public class KolmogorovMinimumWeightPerfectMatching<V, E> implements MatchingAlgorithm<V, E> {
+    /**
+     * Default epsilon used in the algorithm
+     */
+    public static final double EPS = MatchingAlgorithm.DEFAULT_EPSILON;
+    /**
+     * Default infinity value used in the algorithm
+     */
+    public static final double INFINITY = 1e100;
+    /**
+     * Defines the threshold for throwing an exception about no perfect matching existence
+     */
+    public static final double NO_PERFECT_MATCHING_THRESHOLD = 1e10;
+    /**
+     * Variable for debug purposes
+     */
+    static final boolean DEBUG = false;
+    /**
+     * Message about no perfect matching
+     */
+    static final String NO_PERFECT_MATCHING = "There is no perfect matching in the specified graph";
+    /**
+     * Default options
+     */
+    private static final BlossomVOptions DEFAULT_OPTIONS = new BlossomVOptions();
+
+    /**
+     * The graph we are matching on
+     */
+    private final Graph<V, E> graph;
+
+    /**
+     * Current state of the algorithm
+     */
+    BlossomVState<V, E> state;
+    /**
+     * Perform primal operations (grow, augment, shrink and expand)
+     */
+    private BlossomVPrimalUpdater<V, E> primalUpdater;
+    /**
+     * Performs dual updates using the strategy defined by the {@code options}
+     */
+    private BlossomVDualUpdater<V, E> dualUpdater;
+    /**
+     * The computed matching of the {@code graph}
+     */
+    private MatchingAlgorithm.Matching<V, E> matching;
+    /**
+     * Defines solution to the dual linear program formulated on the {@code graph}
+     */
+    private DualSolution dualSolution;
+    /**
+     * BlossomVOptions used by the algorithm to getMatching the problem instance
+     */
+    private BlossomVOptions options;
+
+    /**
+     * Constructs a new instance of the algorithm with the default options for it.
+     *
+     * @param graph the graph a minimum weight perfect matching would be searched in
+     */
+    public KolmogorovMinimumWeightPerfectMatching(Graph<V, E> graph) {
+        this(graph, DEFAULT_OPTIONS);
+    }
+
+    /**
+     * Constructs a new instance of the algorithm with the specified {@code options}
+     *
+     * @param graph   the graph a minimum weight perfect matching would be searched in
+     * @param options the options which define the strategies for the initialization and dual updates
+     */
+    public KolmogorovMinimumWeightPerfectMatching(Graph<V, E> graph, BlossomVOptions options) {
+        Objects.requireNonNull(graph);
+        if ((graph.vertexSet().size() & 1) == 1) {
```

Comment:
I think we should discriminate between easy-to-detect "invalid argument" (like this) and hard-to-detect (such as the cases detected in BlossomVDualUpdater and BlossomVInitializer, which aren't really invalid arguments.

Reply:
Discriminate be throwing an exception of another type, or by mentioning it in the Javadoc?